### PR TITLE
✨Computational backend: connect dv2 to clusters keeper for on-demand clusters (🗃️ + ⚠️devops)

### PR DESCRIPTION
**Pull request by sanderegg**
## What do these changes do?

This PR brings a connection between the director-v2 and the clusters-keeper.
Now a user may have an *on-demand cluster* for his/her computations.
Current limitations are:
- only 1 machine of 1 type

### clusters-keeper
- [x] fixes issue where clusters-keeper would not start in devel mode
- [x] moved RPC interface into models-library
- [x] 🗃️ added column *use_on_demand_clusters* into *comp_runs* table
- [x] new ENV ```CLUSTERS_KEEPER_MAX_MISSED_HEARTBEATS_BEFORE_CLUSTER_TERMINATION``` defaults to 5
- [x] new ENV ```CLUSTERS_KEEPER_TASK_INTERVAL``` defaults to 60
### director-v2
- [x] added exceptions for OnDemandCluster not ready or clusters-keeper not available
- [x] added rabbitMQ RPC client to contact clusters-keeper when on demand cluster is necessary
- [x] added WAITING_FOR_CLUSTER RunningState and in StateType (🗃️ associated)
- [x] added *use_on_demand_clusters* field in ComputationCreate body to instruct dv2 to get or create a cluster on the fly
- [x] currently clusters-keeper is disabled by default in the ENV variables
### frontend
- [x] added WAITING_FOR_CLUSTER RunningState

## How to test

- need access to S3 (AWS or privately available)
- need access to docker registry to run services
- define a GPU oriented machine
- all this goes in a well defined *.env* file that I can provide
- pretty early testing... expect bugs and this is not yet available to all

## DevOps Checklist
- [x] new ENV ```CLUSTERS_KEEPER_MAX_MISSED_HEARTBEATS_BEFORE_CLUSTER_TERMINATION``` defaults to 5
- [x] new ENV ```CLUSTERS_KEEPER_TASK_INTERVAL``` defaults to 60

### CODE REVIEW

File: services/docker-compose.yml [hunk 1]
```diff
@@ -82,16 +82,25 @@ services:
...
-      - REDIS_PORT=${REDIS_PORT}
+      # - CLUSTERS_KEEPER_MAX_MISSED_HEARTBEATS_BEFORE_CLUSTER_TERMINATION=${CLUSTERS_KEEPER_MAX_MISSED_HEARTBEATS_BEFORE_CLUSTER_TERMINATION}
+      # - CLUSTERS_KEEPER_TASK_INTERVAL=${CLUSTERS_KEEPER_TASK_INTERVAL}
+      # - EC2_ACCESS_KEY_ID=${EC2_ACCESS_KEY_ID}
+      # - EC2_INSTANCES_ALLOWED_TYPES=${EC2_INSTANCES_ALLOWED_TYPES}
+      # - EC2_INSTANCES_AMI_ID=${EC2_INSTANCES_AMI_ID}
+      # - EC2_INSTANCES_KEY_NAME=${EC2_INSTANCES_KEY_NAME}
+      # - EC2_INSTANCES_SECURITY_GROUP_IDS=${EC2_INSTANCES_SECURITY_GROUP_IDS}
+      # - EC2_INSTANCES_SUBNET_ID=${EC2_INSTANCES_SUBNET_ID}
+      # - EC2_SECRET_ACCESS_KEY=${EC2_SECRET_ACCESS_KEY}
```
Comment (GitHK) [inline services/docker-compose.yml:94]: did you forget these?

**Review by GitHK — APPROVED** [review]

File: services/docker-compose.yml (hunk 1 above)
Comment (sanderegg) [reply services/docker-compose.yml:94]: no, but I will remove them now. thanks.

File: packages/models-library/src/models_library/api_schemas_directorv2/comp_tasks.py [hunk 2]
```diff
@@ -6,6 +9,8 @@
 from ..projects_pipeline import ComputationTask
 from ..users import UserID
 
+TEMPORARY_DEFAULT_WALLET_ID: Final[WalletID] = 43
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply packages/models-library/src/models_library/api_schemas_directorv2/comp_tasks.py:12]: I will move this away. it is not used for now.

**Review by YuryHrytsuk — APPROVED** [review]

**Comment by sanderegg** [comment]
> Could you in later cases add new env in `DevOps Checklist` so that it is clear and easily seen 🙏

@YuryHrytsuk 
done

File: packages/models-library/src/models_library/api_schemas_directorv2/comp_tasks.py [hunk 3]
```diff
@@ -45,6 +49,14 @@ def ensure_product_name_defined_if_computation_starts(cls, v, values):
             raise ValueError(msg)
         return v
 
+    @validator("use_on_demand_clusters", always=True)
+    @classmethod
+    def ensure_expected_options(cls, v, values):
+        if v is True and ("cluster_id" in values and values["cluster_id"] is not None):
```
Comment (pcrespov) [inline packages/models-library/src/models_library/api_schemas_directorv2/comp_tasks.py:55]: TIP: 
```python
if v is True and value.get("cluster_id") is not None:
    ...
```

File: packages/models-library/src/models_library/api_schemas_directorv2/comp_tasks.py [hunk 4]
```diff
@@ -36,6 +36,10 @@ class ComputationCreate(BaseModel):
         description="the computation shall use the cluster described by its id, 0 is the default cluster",
     )
     simcore_user_agent: str = ""
+    use_on_demand_clusters: bool = Field(
+        default=False,
+        description="if True, a cluster will be created as necessary (wallet_id cannot be None, and cluster_id must be None)",
```
Comment (pcrespov) [inline packages/models-library/src/models_library/api_schemas_directorv2/comp_tasks.py:41]: Q: the validator below do not check for wallet_id!?

File: services/clusters-keeper/src/simcore_service_clusters_keeper/modules/dask.py [hunk 5]
```diff
@@ -22,15 +23,19 @@ async def ping_gateway(*, url: AnyUrl, password: SecretStr) -> bool:
...
+            await asyncio.wait_for(gateway.list_clusters(), timeout=5)
         return True
     except asyncio.TimeoutError:
-        _logger.debug("gateway ping timed-out, it is still starting...")
-    except ClientError:
+        _logger.info(
+            "osparc-gateway %s ping timed-out, the machine is likely still starting...",
+            url,
+        )
+    except (ClientError, ValueError):
```
Comment (pcrespov) [inline services/clusters-keeper/src/simcore_service_clusters_keeper/modules/dask.py:33]: Q: where might be this `ValueError` be raised from? `ValueError` normally denote "wrong arguments" but `list_cluster` has not arguments so `Gateway`?

File: services/clusters-keeper/src/simcore_service_clusters_keeper/rpc/clusters.py [hunk 6]
```diff
@@ -37,17 +39,16 @@ async def get_or_create_cluster(
         ec2_instance = new_ec2_instances[0]
     assert ec2_instance is not None  # nosec
     app_settings = get_application_settings(app)
-    cluster_get = ClusterGet.from_ec2_instance_data(
+    return create_cluster_from_ec2_instance(
         ec2_instance,
         user_id,
         wallet_id,
         app_settings.CLUSTERS_KEEPER_COMPUTATIONAL_BACKEND_GATEWAY_PASSWORD,
+        gateway_ready=bool(
```
Comment (pcrespov) [inline services/clusters-keeper/src/simcore_service_clusters_keeper/rpc/clusters.py:47]: Q: so if `gateway_ready==False` how does the system reponds? i.e. what is the workflow in that situation? do you retry?

File: services/clusters-keeper/src/simcore_service_clusters_keeper/_meta.py [hunk 7]
```diff
@@ -43,12 +43,12 @@ def get_summary() -> str:
  _______  _                 _______ _________ _______  _______  _______         _        _______  _______  _______  _______  _______
 (  ____ \( \      |\     /|(  ____ \\__   __/(  ____ \(  ____ )(  ____ \       | \    /\(  ____ \(  ____ \(  ____ )(  ____ \(  ____ )
 | (    \/| (      | )   ( || (    \/   ) (   | (    \/| (    )|| (    \/       |  \  / /| (    \/| (    \/| (    )|| (    \/| (    )|
-| |      | |      | |   | || (_____    | |   | (__    | (____)|| (_____        |  (_/ / | (__    | (__    | (____)|| (__    | (____)|
-| |      | |      | |   | |(_____  )   | |   |  __)   |     __)(_____  )       |   _ (  |  __)   |  __)   |  _____)|  __)   |     __)
+| |      | |      | |   | || (_____    | |   | (__    | (____)|| (_____  _____ |  (_/ / | (__    | (__    | (____)|| (__    | (____)|
```
Comment (pcrespov) [inline services/clusters-keeper/src/simcore_service_clusters_keeper/_meta.py:46]: sorry, but no taste at all for cool banners ;-b

**Review by pcrespov — APPROVED** [review]
:tada: Cool. Some minor comments/questions

File: packages/models-library/src/models_library/api_schemas_directorv2/comp_tasks.py (hunk 4 above)
Comment (sanderegg) [reply packages/models-library/src/models_library/api_schemas_directorv2/comp_tasks.py:41]: no cause at the moment there are none.

File: services/clusters-keeper/src/simcore_service_clusters_keeper/modules/dask.py (hunk 5 above)
Comment (sanderegg) [reply services/clusters-keeper/src/simcore_service_clusters_keeper/modules/dask.py:33]: ask dask-gateway

File: services/clusters-keeper/src/simcore_service_clusters_keeper/rpc/clusters.py (hunk 6 above)
Comment (sanderegg) [reply services/clusters-keeper/src/simcore_service_clusters_keeper/rpc/clusters.py:47]: it retries again 5 seconds later

**Pull request merged**